scripting/upgrades/smrpg_upgrade_turtle.sp

```diff
@@ -0,0 +1,176 @@
+	{
+		if(IsClientInGame(i))
+			OnClientPutInServer(i);
+	}
+}
+
+public void OnPluginEnd()
+{
+	if(SMRPG_UpgradeExists(UPGRADE_SHORTNAME))
+		SMRPG_UnregisterUpgradeType(UPGRADE_SHORTNAME);
+}
+
+public void OnAllPluginsLoaded()
+{
+	OnLibraryAdded("smrpg");
+}
+
+public void OnLibraryAdded(const char[] name)
+{
+	// Register this upgrade in SM:RPG
+	if(StrEqual(name, "smrpg"))
+	{
+		// Register the upgrade type.
+		SMRPG_RegisterUpgradeType("Turtle", UPGRADE_SHORTNAME, "Reduces the damage you receive after taking a headshot.", 0, true, 5, 10, 10);
+		
+		SMRPG_SetUpgradeTranslationCallback(UPGRADE_SHORTNAME, SMRPG_TranslateUpgrade);
+		
+		// Create your convars through the SM:RPG core. That way they are added to your upgrade's own config file in cfg/sourcemod/smrpg/smrpg_upgrade_example.cfg!
+		g_hCVPercent = SMRPG_CreateUpgradeConVar(UPGRADE_SHORTNAME, "smrpg_turtle_percent", "0.01", "How much percent of the damage should be removed (multiplied by level)?", _, true, 0.01, true, 1.0);
+		g_hCVDuration = SMRPG_CreateUpgradeConVar(UPGRADE_SHORTNAME, "smrpg_turtle_duration", "0.01", "How long should the damage reduction last (multiplied by level)?", _, true, 0.01, true, 10.0);
+	}
+}
+
+public void OnClientPutInServer(int client)
+{
+	SDKHook(client, SDKHook_OnTakeDamage, Hook_OnTakeDamage);
+	SDKHook(client, SDKHook_TraceAttackPost, Hook_TraceAttackPost);
+}
+
+/**
+ * SM:RPG Upgrade callbacks
+ */
+
+// The core wants to display your upgrade somewhere. Translate it into the clients language!
+public void SMRPG_TranslateUpgrade(int client, const char[] shortname, TranslationType type, char[] translation, int maxlen)
+{
+	// Easy pattern is to use the shortname of your upgrade in the translation file
+	if(type == TranslationType_Name)
+		Format(translation, maxlen, "%T", UPGRADE_SHORTNAME, client);
+	// And "shortname description" as phrase in the translation file for the description.
+	else if(type == TranslationType_Description)
+	{
+		char sDescriptionKey[MAX_UPGRADE_SHORTNAME_LENGTH+12] = UPGRADE_SHORTNAME;
+		StrCat(sDescriptionKey, sizeof(sDescriptionKey), " description");
+		Format(translation, maxlen, "%T", sDescriptionKey, client);
+	}
+}
+
+/**
+ * SDK Hooks callbacks
+ */
+
+
+public void Hook_TraceAttackPost(int victim, int attacker, int inflictor, float damage, int damagetype, int ammotype, int hitbox, int hitgroup)
+{
+	// Check if the victim is a player and if the last hit was a headshot.
+    	if (IsClientInGame(victim) && hitgroup == 1 && hitbox == 12 && turtlemode == false) 
+	{
+    // Check if the victim has the "turtle" skill.
+    	int iLevel = SMRPG_GetClientUpgradeLevel(victim, "turtle");
+        if (iLevel > 0) 
+		{
+        // Activate "turtlemode" for the player
+		SetTurtlemode(victim, true); 
+        // Schedule a timer to deactivate "turtlemode" after x seconds 
+		float turtleDuration = float(iLevel) * g_hCVDuration.FloatValue;
+		CreateTimer(turtleDuration, RemoveTurtleMode, victim, TIMER_FLAG_NO_MAPCHANGE);
```

Comment:
Maybe we should delete timer if player would disconnect?

Reply:
Would that be necessary? I thought if it runs out it doesn't matter.